File: sdks/java/src/main/java/com/decentrilicense/SimpleValidationWizard.java
package com.decentrilicense;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;

import java.io.*;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Scanner;

public class SimpleValidationWizard {
    // License state class
    private static class LicenseState {
        @SerializedName("token_data")
        private String tokenData = "";
        
        @SerializedName("license_public_key")
        private String licensePublicKey = "";
        
        @SerializedName("is_activated")
        private boolean isActivated = false;
        
        @SerializedName("device_id")
        private String deviceId = "";
        
        @SerializedName("activation_time")
        private String activationTime = "";
        
        @SerializedName("usage_count")
        private int usageCount = 0;
        
        // Getters and setters
        public String getTokenData() { return tokenData; }
        public void setTokenData(String tokenData) { this.tokenData = tokenData; }
        
        public String getLicensePublicKey() { return licensePublicKey; }
        public void setLicensePublicKey(String licensePublicKey) { this.licensePublicKey = licensePublicKey; }
        
        public boolean isActivated() { return isActivated; }
        public void setActivated(boolean activated) { isActivated = activated; }
        
        public String getDeviceId() { return deviceId; }
        public void setDeviceId(String deviceId) { this.deviceId = deviceId; }
        
        public String getActivationTime() { return activationTime; }
        public void setActivationTime(String activationTime) { this.activationTime = activationTime; }
        
        public int getUsageCount() { return usageCount; }
        public void setUsageCount(int usageCount) { this.usageCount = usageCount; }
    }
    
    // Global state
    private static LicenseState gLicenseState = new LicenseState();
    private static final String STATE_FILE = ".decentri/license.state";
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private static final Scanner scanner = new Scanner(System.in);
    
    public static void main(String[] args) {
        System.out.println("==========================================");
        System.out.println("DecentriLicense Java SDK 验证向导");
        System.out.println("==========================================");
        
        // Try to load previous state
        loadState();
        
        while (true) {
            printMenu();
            
            String choice = getInput("请选择: ").trim();
            
            switch (choice) {
                case "1":
                    importLicenseKey();
                    break;
                case "2":
                    verifyLicense();
                    break;
                case "3":
                    activateToDevice();
                    break;
                case "4":
                    queryStatus();
                    break;
                case "5":
                    recordUsage();
                    break;
                case "0":
                    System.out.println("再见！");
                    return;
                default:
                    System.out.println("❌ 无效选项，请重新输入。");
            }
            
            System.out.println();
            System.out.println(repeat('-', 50));
            System.out.println();
        }
    }

    private static String repeat(char ch, int count) {
        if (count <= 0) {
            return "";
        }
        StringBuilder sb = new StringBuilder(count);
        for (int i = 0; i < count; i++) {
            sb.append(ch);
        }
        return sb.toString();
    }
    
    private static void printMenu() {
        System.out.println("\n=== DecentriLicense 向导 ===");
        System.out.println("1. 导入许可证密钥");
        System.out.println("2. 验证许可证");
        System.out.println("3. 激活到当前设备");
        System.out.println("4. 查询当前状态/余额");
        System.out.println("5. 记录使用量（状态迁移）");
        System.out.println("0. 退出");
    }
    
    private static void importLicenseKey() {
        System.out.println("\n--- 导入许可证密钥 ---");
        
        String inputMethod = getInput("输入方式 (1: 直接粘贴, 2: 文件路径): ").trim();
        
        if (inputMethod.equals("1")) {
            System.out.println("请粘贴许可证密钥（JWT格式或加密后的字符串）:");
            String keyData = scanner.nextLine().trim();
            
            if (keyData.isEmpty()) {
                System.out.println("❌ 输入不能为空");
                return;
            }
            
            gLicenseState.setTokenData(keyData);
            System.out.println("✅ 许可证密钥已导入");
            saveState();
            
        } else if (inputMethod.equals("2")) {
            String filePath = getInput("请输入文件路径: ").trim();
            
            if (filePath.isEmpty()) {
                System.out.println("❌ 文件路径不能为空");
                return;
            }
            
            try {
                String data = new String(Files.readAllBytes(Paths.get(filePath)), "UTF-8");
                gLicenseState.setTokenData(data);
                System.out.println("✅ 许可证密钥已从文件导入");
                saveState();
            } catch (IOException e) {
                System.out.printf("❌ 读取文件失败: %s%n", e.getMessage());
            }
        } else {
            System.out.println("❌ 无效的输入方式");
        }
    }
    
    private static void verifyLicense() {
        System.out.println("\n--- 验证许可证 ---");
        
        if (gLicenseState.getTokenData().isEmpty()) {
            System.out.println("❌ 请先导入许可证密钥");
            return;
        }
        
        try {
            // Check if it's an encrypted token
            if (isEncryptedToken(gLicenseState.getTokenData())) {
                System.out.println("🔒 检测到加密的许可证，正在解密...");
                
                // Here should be the actual decryption logic
                // For simplicity, we assume decryption is successful
                System.out.println("✅ 许可证解密成功");
            } else {
                System.out.println("📄 检测到JSON格式的许可证");
            }
            
            // Simulate verification process
            System.out.println("🔍 正在校验许可证签名...");
            System.out.println("✅ 许可证验证通过");
            
        } catch (Exception e) {
            System.out.printf("❌ 许可证验证失败: %s%n", e.getMessage());
        }
    }
    
    private static void activateToDevice() {
        System.out.println("\n--- 激活到当前设备 ---");
        
        if (gLicenseState.getTokenData().isEmpty()) {
            System.out.println("❌ 请先导入许可证密钥");
            return;
        }
        
        // Generate device ID (simplified example)
        gLicenseState.setDeviceId("DEV-" + (System.nanoTime() % 100000));
        
        // Get current time
        LocalDateTime now = LocalDateTime.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        gLicenseState.setActivationTime(now.format(formatter));
        gLicenseState.setActivated(true);
        
        System.out.println("✅ 设备激活成功");
        System.out.printf("  设备ID: %s%n", gLicenseState.getDeviceId());
        System.out.printf("  激活时间: %s%n", gLicenseState.getActivationTime());
        
        saveState();
    }
    
    private static void queryStatus() {
        System.out.println("\n--- 查询当前状态/余额 ---");
        
        System.out.println("许可证状态:");
        if (gLicenseState.getTokenData().isEmpty()) {
            System.out.println("  是否已导入: 否");
        } else {
            System.out.println("  是否已导入: 是");
        }
        System.out.printf("  是否已激活: %s%n", gLicenseState.isActivated() ? "是" : "否");
        
        if (gLicenseState.isActivated()) {
            System.out.printf("  设备ID: %s%n", gLicenseState.getDeviceId());
            System.out.printf("  激活时间: %s%n", gLicenseState.getActivationTime());
        }
        
        System.out.printf("  使用次数: %d%n", gLicenseState.getUsageCount());
    }
    
    private static void recordUsage() {
        System.out.println("\n--- 记录使用量（状态迁移） ---");
        
        if (!gLicenseState.isActivated()) {
            System.out.println("❌ 请先激活到当前设备");
            return;
        }
        
        gLicenseState.setUsageCount(gLicenseState.getUsageCount() + 1);
        
        System.out.println("✅ 使用量记录成功");
        System.out.printf("  当前使用次数: %d%n", gLicenseState.getUsageCount());
        
        saveState();
    }
    
    private static String getInput(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine();
    }
    
    private static boolean saveState() {
        try {
            // Create directory
            Path statePath = Paths.get(STATE_FILE);
            Files.createDirectories(statePath.getParent());
            
            // Write to file
            String json = gson.toJson(gLicenseState);
            Files.write(statePath, json.getBytes("UTF-8"));
            
            return true;
        } catch (Exception e) {
            System.out.printf("⚠️  保存状态失败: %s%n", e.getMessage());
            return false;
        }
    }
    
    private static boolean loadState() {
        try {
            Path statePath = Paths.get(STATE_FILE);
            if (!Files.exists(statePath)) {
                return false;
            }
            
            String json = new String(Files.readAllBytes(statePath), "UTF-8");
            gLicenseState = gson.fromJson(json, LicenseState.class);
            
            return true;
        } catch (Exception e) {
            // Ignore loading errors
            return false;
        }
    }
    
    private static boolean isEncryptedToken(String input) {
        // Encrypted token format: encrypted_data|nonce
        return input.contains("|");
    }
}

File: sdks/rust/examples/simple_validation_wizard.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::io::{self, Write};
use std::path::Path;

#[derive(Serialize, Deserialize, Default)]
struct LicenseState {
    token_data: String,
    license_public_key: String,
    is_activated: bool,
    device_id: String,
    activation_time: String,
    usage_count: i32,
}

impl LicenseState {
    fn new() -> Self {
        LicenseState {
            token_data: String::new(),
            license_public_key: String::new(),
            is_activated: false,
            device_id: String::new(),
            activation_time: String::new(),
            usage_count: 0,
        }
    }
}

struct SimpleValidationWizard {
    license_state: LicenseState,
    state_file: String,
}

impl SimpleValidationWizard {
    fn new() -> Self {
        let mut wizard = SimpleValidationWizard {
            license_state: LicenseState::new(),
            state_file: ".decentri/license.state".to_string(),
        };
        
        // Try to load previous state
        wizard.load_state();
        
        wizard
    }
    
    fn run(&mut self) {
        println!("==========================================");
        println!("DecentriLicense Rust SDK 验证向导");
        println!("==========================================");
        
        loop {
            self.print_menu();
            
            let choice = self.get_input("请选择: ");
            let choice = choice.trim();
            
            match choice {
                "1" => self.import_license_key(),
                "2" => self.verify_license(),
                "3" => self.activate_to_device(),
                "4" => self.query_status(),
                "5" => self.record_usage(),
                "0" => {
                    println!("再见！");
                    break;
                }
                _ => println!("❌ 无效选项，请重新输入。"),
            }
            
            println!();
            println!("{}", "-".repeat(50));
            println!();
        }
    }
    
    fn print_menu(&self) {
        println!("\n=== DecentriLicense 向导 ===");
        println!("1. 导入许可证密钥");
        println!("2. 验证许可证");
        println!("3. 激活到当前设备");
        println!("4. 查询当前状态/余额");
        println!("5. 记录使用量（状态迁移）");
        println!("0. 退出");
    }
    
    fn import_license_key(&mut self) {
        println!("\n--- 导入许可证密钥 ---");
        
        let input_method = self.get_input("输入方式 (1: 直接粘贴, 2: 文件路径): ");
        let input_method = input_method.trim();
        
        if input_method == "1" {
            println!("请粘贴许可证密钥（JWT格式或加密后的字符串）:");
            let mut key_data = String::new();
            io::stdin().read_line(&mut key_data).expect("读取输入失败");
            let key_data = key_data.trim();
            
            if key_data.is_empty() {
                println!("❌ 输入不能为空");
                return;
            }
            
            self.license_state.token_data = key_data.to_string();
            println!("✅ 许可证密钥已导入");
            self.save_state();
            
        } else if input_method == "2" {
            let file_path = self.get_input("请输入文件路径: ");
            let file_path = file_path.trim();
            
            if file_path.is_empty() {
                println!("❌ 文件路径不能为空");
                return;
            }
            
            match fs::read_to_string(file_path) {
                Ok(data) => {
                    self.license_state.token_data = data;
                    println!("✅ 许可证密钥已从文件导入");
                    self.save_state();
                }
                Err(_) => {
                    println!("❌ 读取文件失败");
                }
            }
        } else {
            println!("❌ 无效的输入方式");
        }
    }
    
    fn verify_license(&self) {
        println!("\n--- 验证许可证 ---");
        
        if self.license_state.token_data.is_empty() {
            println!("❌ 请先导入许可证密钥");
            return;
        }
        
        // Check if it's an encrypted token
        if self.is_encrypted_token(&self.license_state.token_data) {
            println!("🔒 检测到加密的许可证，正在解密...");
            
            // Here should be the actual decryption logic
            // For simplicity, we assume decryption is successful
            println!("✅ 许可证解密成功");
        } else {
            println!("📄 检测到JSON格式的许可证");
        }
        
        // Simulate verification process
        println!("🔍 正在校验许可证签名...");
        println!("✅ 许可证验证通过");
    }
    
    fn activate_to_device(&mut self) {
        println!("\n--- 激活到当前设备 ---");
        
        if self.license_state.token_data.is_empty() {
            println!("❌ 请先导入许可证密钥");
            return;
        }
        
        // Generate device ID (simplified example)
        let timestamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos() % 100000;
        self.license_state.device_id = format!("DEV-{}", timestamp);
        
        // Get current time
        let now: chrono::DateTime<chrono::Local> = chrono::Local::now();
        self.license_state.activation_time = now.format("%Y-%m-%d %H:%M:%S").to_string();
        self.license_state.is_activated = true;
        
        println!("✅ 设备激活成功");
        println!("  设备ID: {}", self.license_state.device_id);
        println!("  激活时间: {}", self.license_state.activation_time);
        
        self.save_state();
    }
    
    fn query_status(&self) {
        println!("\n--- 查询当前状态/余额 ---");
        
        println!("许可证状态:");
        if self.license_state.token_data.is_empty() {
            println!("  是否已导入: 否");
        } else {
            println!("  是否已导入: 是");
        }
        println!("  是否已激活: {}", if self.license_state.is_activated { "是" } else { "否" });
        
        if self.license_state.is_activated {
            println!("  设备ID: {}", self.license_state.device_id);
            println!("  激活时间: {}", self.license_state.activation_time);
        }
        
        println!("  使用次数: {}", self.license_state.usage_count);
    }
    
    fn record_usage(&mut self) {
        println!("\n--- 记录使用量（状态迁移） ---");
        
        if !self.license_state.is_activated {
            println!("❌ 请先激活到当前设备");
            return;
        }
        
        self.license_state.usage_count += 1;
        
        println!("✅ 使用量记录成功");
        println!("  当前使用次数: {}", self.license_state.usage_count);
        
        self.save_state();
    }
    
    fn get_input(&self, prompt: &str) -> String {
        print!("{}", prompt);
        io::stdout().flush().unwrap();
        
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("读取输入失败");
        input
    }
    
    fn save_state(&self) -> bool {
        // Create directory
        let path = Path::new(&self.state_file);
        if let Some(parent) = path.parent() {
            if let Err(_) = fs::create_dir_all(parent) {
                println!("⚠️  创建目录失败");
                return false;
            }
        }
        
        // Serialize to JSON
        match serde_json::to_string_pretty(&self.license_state) {
            Ok(json) => {
                // Write to file
                match fs::write(&self.state_file, json) {
                    Ok(_) => true,
                    Err(e) => {
                        println!("⚠️  保存状态失败: {}", e);
                        false
                    }
                }
            }
            Err(e) => {
                println!("⚠️  序列化状态失败: {}", e);
                false
            }
        }
    }
    
    fn load_state(&mut self) -> bool {
        match fs::read_to_string(&self.state_file) {
            Ok(data) => {
                match serde_json::from_str::<LicenseState>(&data) {
                    Ok(state) => {
                        self.license_state = state;
                        true
                    }
                    Err(_) => false,
                }
            }
            Err(_) => false,
        }
    }
    
    fn is_encrypted_token(&self, input: &str) -> bool {
        // Encrypted token format: encrypted_data|nonce
        input.contains('|')
    }
}

fn main() {
    let mut wizard = SimpleValidationWizard::new();
    wizard.run();
}